### ISSUE TITLE
[x86] Backend crash during pass `X86 FP Stackifier`

### ISSUE THREAD

**Issue by HazyFish** [llvm/llvm-project#58924]
## Description

When targeting `i386` or `i686`, the following code crashes the backend during pass `X86 FP Stackifier`.

The problem doesn't exist when targeting `x86_64` or `aarch64`.
The problem also doesn't exist when `v4f64` instead of `v8f64` is used.

## Minimal Reproduction

https://godbolt.org/z/1vh8jf7YT

### Code 

```ll
; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare float @llvm.pow.f32(float, float) #0

define float @f(<8 x double> %0, i1 %1, <8 x double>* %2, float* %3) {
BB:
  %C2 = call float @llvm.pow.f32(float 0.25, float 0.5)
  %C = fcmp true double 0.25, 0.5
  %B1 = fmul <8 x double> %0, <double 0.25, double 0.25, double 0.25, double 0.25, double 0.25, double 0.25, double 0.25, double 0.25>
  %B = or i1 %1, %C
  %F = load float, float* %3
  store <8 x double> %B1, <8 x double>* %2
  br i1 %B, label %BB1, label %BB2

BB1:                                             ; preds = %BB
  ret float %F

BB2:                                             ; preds = %BB
  ret float %C2
}
```

### Stack Trace

```console
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: ./llvm-project/build-debug/bin/llc -mtriple=i686 ./crash-reports/dagisel-i686/1.ll
1.	Running pass 'Function Pass Manager' on module './crash-reports/dagisel-i686/1.ll'.
2.	Running pass 'X86 FP Stackifier' on function '@f'
 #0 0x00000000047ada8a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:569:11
 #1 0x00000000047adc3b PrintStackTraceSignalHandler(void*) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:636:1
 #2 0x00000000047ac286 llvm::sys::RunSignalHandlers() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Signals.cpp:104:5
 #3 0x00000000047ae365 SignalHandler(int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f7a1c8fa980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #5 0x0000000002614df6 (anonymous namespace)::FPS::freeStackSlotBefore(llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>, unsigned int) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/X86/X86FloatingPoint.cpp:907:21
 #6 0x000000000261475c (anonymous namespace)::FPS::adjustLiveRegs(unsigned int, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/X86/X86FloatingPoint.cpp:961:5
 #7 0x0000000002613fea (anonymous namespace)::FPS::finishBlockStack() /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/X86/X86FloatingPoint.cpp:558:8
 #8 0x0000000002610c56 (anonymous namespace)::FPS::processBasicBlock(llvm::MachineFunction&, llvm::MachineBasicBlock&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/X86/X86FloatingPoint.cpp:501:10
 #9 0x000000000260fed8 (anonymous namespace)::FPS::runOnMachineFunction(llvm::MachineFunction&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/Target/X86/X86FloatingPoint.cpp:375:16
#10 0x00000000035783e5 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:91:8
#11 0x0000000003c55c66 llvm::FPPassManager::runOnFunction(llvm::Function&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1430:23
#12 0x0000000003c5aa92 llvm::FPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1476:16
#13 0x0000000003c56539 (anonymous namespace)::MPPassManager::runOnModule(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1545:23
#14 0x0000000003c560ad llvm::legacy::PassManagerImpl::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:535:16
#15 0x0000000003c5ad71 llvm::legacy::PassManager::run(llvm::Module&) /home/henry/aflplusplus-isel/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1672:3
#16 0x0000000000d38a5c compileModule(char**, llvm::LLVMContext&) /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:737:41
#17 0x0000000000d36e02 main /home/henry/aflplusplus-isel/llvm-project/llvm/tools/llc/llc.cpp:418:13
#18 0x00007f7a1b7cdc87 __libc_start_main /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:344:0
#19 0x0000000000d3660a _start (./llvm-project/build-debug/bin/llc+0xd3660a)
```

**Comment by topperc** [comment]
The pow result is being detected as dead way later than we expect. We end up with a KILL instruction for fp0 which isn't handled in the stackifier.

Here's a possible patch

```
diff --git a/llvm/lib/Target/X86/X86FloatingPoint.cpp b/llvm/lib/Target/X86/X86FloatingPoint.cpp
index 33f5bb365da8..9a18838551e0 100644
--- a/llvm/lib/Target/X86/X86FloatingPoint.cpp
+++ b/llvm/lib/Target/X86/X86FloatingPoint.cpp
@@ -426,6 +426,10 @@ bool FPS::processBasicBlock(MachineFunction &MF, MachineBasicBlock &BB) {
     if (MI.isCopy() && isFPCopy(MI))
       FPInstClass = X86II::SpecialFP;
 
+    if (MI.isKill() && MI.getNumOperands() > 0 &&
+        X86::RFP80RegClass.contains(MI.getOperand(0).getReg()))
+      FPInstClass = X86II::SpecialFP;
+
     if (MI.isImplicitDef() &&
         X86::RFP80RegClass.contains(MI.getOperand(0).getReg()))
       FPInstClass = X86II::SpecialFP;
@@ -450,6 +454,10 @@ bool FPS::processBasicBlock(MachineFunction &MF, MachineBasicBlock &BB) {
       if (MO.isReg() && MO.isDead())
         DeadRegs.push_back(MO.getReg());
 
+    if (MI.isKill() && MI.getNumOperands() > 0 &&
+        X86::RFP80RegClass.contains(MI.getOperand(0).getReg()))
+      DeadRegs.push_back(MI.getOperand(0).getReg());
+
     switch (FPInstClass) {
     case X86II::ZeroArgFP:  handleZeroArgFP(I); break;
     case X86II::OneArgFP:   handleOneArgFP(I);  break;  // fstp ST(0)
@@ -1522,6 +1530,10 @@ void FPS::handleSpecialFP(MachineBasicBlock::iterator &Inst) {
 
   switch (MI.getOpcode()) {
   default: llvm_unreachable("Unknown SpecialFP instruction!");
+  case TargetOpcode::KILL:
+    // Insert a dummy kill.
+    BuildMI(*MBB, Inst, DebugLoc(), TII->get(TargetOpcode::KILL));
+    break;
   case TargetOpcode::COPY: {
     // We handle three kinds of copies: FP <- FP, FP <- ST, and ST <- FP.
     const MachineOperand &MO1 = MI.getOperand(1);
```

**Comment by DataCorrupted** [comment]
Hi Craig, thanks for the explanation and the patch! Really helpful!
Since I didn't see the diff in Phabricator, have you send it there? Or would you like our help to do it?

**Comment by DataCorrupted** [comment]
Fixed as of today, still crashing in LLVM 17.0.1
